clippy_lints/src/format_args.rs

```diff
@@ -559,6 +605,59 @@ impl<'tcx> FormatArgsExpr<'_, 'tcx> {
 
         false
     }
+
+    fn has_pointer_debug(&mut self, ty: Ty<'tcx>) -> bool {
+        let mut visited = FxHashSet::default();
+        let mut open: Vec<Ty<'_>> = Vec::new();
+        open.push(ty);
+        let mut contains = Vec::new();
+        while let Some(ty) = open.pop() {
+            match ty.kind() {
+                ty::RawPtr(..) | ty::FnPtr(..) | ty::FnDef(..) => return true,
+                ty::Ref(_, t, _) | ty::Slice(t) | ty::Array(t, _) => open.push(*t),
+                ty::Tuple(ts) => open.extend_from_slice(ts),
+                ty::Adt(adt, args) => {
+                    // avoid infinite recursion
+                    if !visited.insert(adt.did()) {
+                        continue;
+                    }
+                    let cx = self.cx;
+                    let tcx = cx.tcx;
+                    let has_derived_debug = if let Some(known) = self.has_derived_debug.get(&ty) {
+                        *known
+                    } else {
+                        let Some(trait_id) = tcx.get_diagnostic_item(sym::Debug) else {
+                            continue;
+                        };
+                        let (infcx, param_env) = tcx.infer_ctxt().build_with_typing_env(cx.typing_env());
+                        let ty = tcx.erase_regions(ty);
+                        let trait_ref = TraitRef::new(tcx, trait_id, [GenericArg::from(ty)]);
+                        let obligation = Obligation {
+                            cause: ObligationCause::dummy(),
+                            param_env,
+                            recursion_depth: 0,
+                            predicate: trait_ref.upcast(tcx),
+                        };
+                        let selection = SelectionContext::new(&infcx).select(&obligation);
+                        let Ok(Some(Selection::UserDefined(data))) = selection else {
+                            continue;
+                        };
+                        let has_derived_debug = tcx.has_attr(data.impl_def_id, sym::automatically_derived);
+                        self.has_derived_debug.insert(ty, has_derived_debug);
+                        has_derived_debug
+                    };
+                    // we currently only look into derived impls because those will
+                    // debug-format the types fields which is easy enough to pull off
+                    if has_derived_debug {
+                        contains.push(ty);
+                        open.extend(adt.all_fields().map(|f| f.ty(tcx, args)));
```

Comment:
These types all need to be normalized via `TyCtxt::try_normalize_erasing_regions`.